core/src/main/java/com/graphhopper/routing/weighting/DefaultTurnCostProvider.java

```diff
@@ -0,0 +1,92 @@
+import com.graphhopper.storage.IntsRef;
+import com.graphhopper.storage.TurnCostStorage;
+import com.graphhopper.util.EdgeIterator;
+
+import static com.graphhopper.routing.profiles.TurnCost.EV_SUFFIX;
+import static com.graphhopper.routing.util.EncodingManager.getKey;
+import static com.graphhopper.routing.weighting.Weighting.INFINITE_U_TURN_COSTS;
+
+public class DefaultTurnCostProvider implements TurnCostProvider {
+    private final DecimalEncodedValue turnCostEnc;
+    private final TurnCostStorage turnCostStorage;
+    private final int uTurnCostsInt;
+    private final double uTurnCosts;
+    private final IntsRef tcFlags = TurnCost.createFlags();
+
+    public DefaultTurnCostProvider(FlagEncoder encoder, TurnCostStorage turnCostStorage) {
+        this(encoder, turnCostStorage, Weighting.INFINITE_U_TURN_COSTS);
+    }
+
+    /**
+     * @param uTurnCosts the costs of a u-turn in seconds, for {@link Weighting#INFINITE_U_TURN_COSTS} the u-turn costs
+     *                   will be infinite
+     */
+    public DefaultTurnCostProvider(FlagEncoder encoder, TurnCostStorage turnCostStorage, int uTurnCosts) {
+        if (uTurnCosts < 0 && uTurnCosts != INFINITE_U_TURN_COSTS) {
+            throw new IllegalArgumentException("u-turn costs must be positive, or equal to " + INFINITE_U_TURN_COSTS + " (=infinite costs)");
+        }
+        this.uTurnCostsInt = uTurnCosts;
+        this.uTurnCosts = uTurnCosts < 0 ? Double.POSITIVE_INFINITY : uTurnCosts;
+        if (turnCostStorage == null) {
+            throw new IllegalArgumentException("No storage set to calculate turn weight");
+        }
+        String key = getKey(encoder.toString(), EV_SUFFIX);
+        // if null the TurnCostProvider can be still useful for edge-based routing
+        this.turnCostEnc = encoder.hasEncodedValue(key) ? encoder.getDecimalEncodedValue(key) : null;
+        this.turnCostStorage = turnCostStorage;
+    }
+
+    @Override
+    public double calcTurnWeight(int edgeFrom, int nodeVia, int edgeTo) {
+        if (!EdgeIterator.Edge.isValid(edgeFrom) || !EdgeIterator.Edge.isValid(edgeTo)) {
+            return 0;
+        }
+        double tCost = 0;
+        if (turnCostStorage.isUTurn(edgeFrom, edgeTo)) {
+            // note that the u-turn costs overwrite any turn costs set in TurnCostStorage
+            tCost = turnCostStorage.isUTurnAllowed(nodeVia) ? uTurnCosts : Double.POSITIVE_INFINITY;
+        } else {
+            if (turnCostEnc != null)
+                tCost = turnCostStorage.get(turnCostEnc, tcFlags, edgeFrom, nodeVia, edgeTo);
+        }
+        return tCost;
+    }
+
+    @Override
+    public long calcTurnMillis(int inEdge, int viaNode, int outEdge) {
+        return (long) (1000 * calcTurnWeight(inEdge, viaNode, outEdge));
+    }
+
+    @Override
+    public String getName() {
+        return toString();
+    }
+
+    @Override
+    public String toString() {
+        return "" + uTurnCostsInt;
```

Comment:
Shouldn't it include the FlagEncoder too? Because without it, then different providers will have the same name.

(and in weighting.matches we could adapt the equals)

Reply:
Hmm, the `FlagEncoder` belongs to the `Weighting` and the `DefaultTurnCostProvider` in a way.. Right now its included in the `Weighting` name only.

Reply:
I see, this would be a bit of a circular dependency then. Still it feels strange that there will be different turn cost providers with the same name as soon as multiple vehicles are imported.

Reply:
Hmm yes this is not ideal. `DefaultTurnCostProvider` and `AbstractWeighting` are both using a `FlagEncoder` (and it most likely always has to be the same), so maybe the most correct way to assign the names would be using the `FlagEncoder` name for both? But then the weighting name would be something like `car|fastest|car|40`, which seems bad as well. 

> here will be different turn cost providers with the same name 

Maybe the question here is a turn cost provider is something that can exist without a weighting?